File: Array/Medium/findMajority.py
'''
Find the Majority Element that occurs more than N/2 times
'''

'''
using hashing
Time Complexity: O(N log N)
Space Complexity: O(N) as using dictionary for storing
'''
def findMajority(arr):
    bound = len(arr) / 2
    tempDict = {}
    for i in range(0,len(arr)):
        if arr[i] in tempDict:
            tempDict[arr[i]] = tempDict[arr[i]] + 1
        else:
            tempDict[arr[i]] = 1
        if(tempDict[arr[i]] > bound):
            return arr[i]

# arr = [2,2,1,1,1,2,2]
# res = findMajority(arr)
# print(res)


'''
using moores voting algorithm
Time Complexity : 
Space Complexity: 
'''

def MVA(arr):
    count = 0
    ele = 0
    for i in range(len(arr)):
        if count == 0:
            ele = arr[i]
            count = 1
        elif ele == arr[i]:
            count = count + 1
        else:
            count = count -1 
    return ele

arrr = [7,7,5,7,5,4,4,5,6,7,7,5,3,5,7,45,5,5,3,5,5,55,5]
ele = MVA(arrr)
print(ele)

File: Array/Medium/js/findMajority.js
/** Find the Majority Element that occurs more than N/2 times 
 * !Moores Vorting Algorithm
*/

arr = [1, 1, 1,1,1,1, 1,1,1, 3, 3, 5, 6, 8, 4, 3, 8, 1]

/** 
 * Brute Force: Naives Approach
 * TC: O(N) * (O2N) = O(N^2)
 * SC: O(N) if we use done, removing it will reduce the space complexity by O(1), TC remains O(N^2)
 *  */
function findMajortiy(arr) {
    let N = arr.length
    let done = []
    for (let i = 0; i < N; i++) {
        if (!done.includes(arr[i])) {
            done.push(arr[i])
            let cnt = 0
            for (let j = 0; j < N; j++) {
                if (arr[i] == arr[j]) {
                    cnt++
                }
            }
            if (cnt > N / 2) {
                return arr[i]
            }
        }
    }
}
console.log('BruteForce: The element that occurs more then N/2 times is :', findMajortiy(arr))


/** 
 * Using Hashing
 * TC: =~ O(2N)
 * SC: O(N)
 *  */
function findMajorityHash(arr) {
    let hashMap = {}
    let N = arr.length
    let majority
    for (ele of arr) {
        if (hashMap[ele]) {
            hashMap[ele]++
        } else {
            hashMap[ele] = 1
        }
    }
    console.log(hashMap)
    Object.keys(hashMap).forEach(ele => {
        if (hashMap[ele] > (N / 2)) {
            majority = ele
        }
    })
    return majority
}
console.log('Hashing: The element that occurs more then N/2 times is :', findMajorityHash(arr))

/** 
 * Using Moores Voting Algorithm
 * TC: O(N)
 * SC: O(1)
 *  */
function findMajority(arr) {
    let cnt = 0
    let curr
    for (let ele of arr) {
        if (cnt == 0) {
            curr = ele
            cnt++
        } else if (curr == ele) {
            cnt++
        } else {
            cnt--
        }
        return ele
    }
}
console.log('MVA: The element that occurs more then N/2 times is :', findMajortiy(arr))